source/clear-linux/guides/network/firewalls.rst

```diff
@@ -0,0 +1,198 @@
+   concepts and configuration options.
+
+   Below is a common restrictive firewall configuration which denies all
+   incoming connections, unless the connection was initiated by the host. 
+
+   .. code:: bash
+
+      #Delete all existing rules 
+      sudo iptables -F
+
+      # Set default chain policies 
+      sudo iptables -P INPUT DROP
+      sudo iptables -P FORWARD DROP
+      sudo iptables -P OUTPUT ACCEPT
+
+      # Accept on localhost loopback device
+      sudo iptables -A INPUT -i lo -j ACCEPT
+      sudo iptables -A OUTPUT -o lo -j ACCEPT
+
+      # Allow established sessions to receive traffic
+      sudo iptables -A INPUT -m conntrack --ctstate ESTABLISHED,RELATED -j ACCEPT
+
+
+#. Test the running firewall configuration to make sure it behaving as
+   you expect. 
+
+#. Run the *iptables-save* service to make the running configuration
+   persistent. This will perform a one-time save of the running configuration
+   to :file:`/etc/iptables.rules` :
+
+   .. code:: bash
+
+      sudo systemctl start iptables-save
+
+
+
+
+#. Enable the *iptables-resolve* service the iptables rules to be
+   automatically applied at boot from the :file:`/etc/iptables.rules` file:
+
+   .. code :: bash
+
+      sudo systemctl enable iptables-restore.service
+
+
+ipset
+=====
+
+`ipset <http://ipset.netfilter.org/>`_ is a framework in the Linux kernel for
+storing and efficiently indexing combinations of IP addresses, networks,
+(TCP/UDP) port numbers, MAC addresses, and interface names.
+
+Custom ipsets do not need to be defined to use iptables but, iptables can be
+extended with IP sets for matching against a large and/or changing sets of
+hosts.
+
+* Running IP sets can be manipulated with the :command:`ipset` utility. 
+
+* Custom IP sets can be stored in the :file:`/etc/ipset.conf` file and
+  automatically applied at boot by enabling the *ipset* service
+  :command:`systemctl enable ipset`.
+
+See :command:`man ipset` to learn more about using ipsets. 
+
+
+firewalld
+*********
+
+`firewalld <https://firewalld.org/>`_ is based on nftables, the successor to
+iptables and parts of the netfilter framework. 
+
+The description firewalld helps highlight some of the differences compared to
+iptables:
+
+    firewalld provides a dynamically managed firewall with support for
+    network/firewall zones to define the trust level of network connections or
+    interfaces. It has support for IPv4, IPv6 firewall settings and for
+    ethernet bridges and has a separation of runtime and permanent
+    configuration options. It also supports an interface for services or
+    applications to add firewall rules directly.
+
+See :command:`man firewalld` for more information. 
+
+Below is information on using :command:`firewalld` on |CL|: 
+
+#. Make sure the *firewalld* bundle is installed:
+
+   .. code:: bash
+
+      sudo swupd bundle-add firewalld
+
+
+#. Disable *iptables*, *ipset*, and *tallow* services as they conflict with
```

Comment:
Technically not entirely correct. Better would be to modify `tallow.service` to make it *not* create iptables rules - it has an option for that, and have firewalld insert this rule. `man tallow` should document this.

Reply:
I created https://github.com/clearlinux/tallow/issues/11 to add firewalld support to tallow. 

> and have firewalld insert this rule. man tallow should document this.

I'm uncertain if the `firewalld.ipset` firewalld uses is the same underlying structure as the regular `ipset`. https://firewalld.org/documentation/man-pages/firewalld.ipset.html . So even if the firewalld rule is manually created, I don't think tallow would be updating the same data structure.

```
 $ sudo firewall-cmd --get-ipsets
```

```
$ sudo ipset list
Name: tallow
Type: hash:ip
Revision: 4
Header: family inet hashsize 1024 maxelem 65536 timeout 3600
Size in memory: 88
References: 0
Number of entries: 0
Members:

Name: tallow6
Type: hash:ip
Revision: 4
Header: family inet6 hashsize 1024 maxelem 65536 timeout 3600
Size in memory: 104
References: 0
Number of entries: 0
Members:
```

So my understanding is that in this configuration tallows list wouldn't be used anyways, hence I disabled it. Can you confirm my understanding?

Reply:
if `ipset` doesn't show firewalld sets then something is really wrong. I don't know what to do in that case - ipsets can't be "temporarily missing" because iptables will not allow you to insert rules referring to it - because it does not exist.

Reply:
So, in its current state doesn't it make sense to disable tallow if firewalld is chosen for use? 
tallow will be creating ipsets/rules that firewalld cannot not use. 

The net result is the same whether tallow is disabled or if is configured not to create rules, right?

Reply:
Disabling tallow never makes sense, imho. We should absolutely drive a solution that doesn't require users to `if (want_feature_x) { disable(feature_y) }`.

If firewalld is stubborn about it, we should patch it to not break tallow (functionally). Especially if firewalld doesn't appear to be able to behave.

Maybe we need to pre-configure firewalld so it runs *after* tallow and somehow happily lives along with tallow.



> The net result is the same whether tallow is disabled or if is configured not to create rules, right?

Not even close? If tallow doesn't make rules, but -say- iptables-restore.service does, tallow is (essentially) fully functional. If you disable tallow, it is entirely non-functional.

Reply:
> If tallow doesn't make rules, but -say- iptables-restore.service does, tallow is (essentially) fully functional.

But the suggestion here is to mask iptables-restore if firewalld is being used so this case wouldn't occur.

In any case, I'll update the document to create a tallow.conf with `nocreate` instead.